vibrato/src/feature_extractor.rs

```diff
@@ -0,0 +1,362 @@
+            });
+        }
+
+        let mut left_parsed_templates = vec![];
+        let mut right_parsed_templates = vec![];
+        for (left_template, right_template) in bigram_templates {
+            {
+                let raw_template = left_template.to_string();
+                let mut required_indices = vec![];
+                let mut captures = vec![];
+                for m in left_feature_pattern.captures_iter(&raw_template) {
+                    let pattern = m.get(0).unwrap();
+                    let idx: usize = m.get(2).unwrap().as_str().parse().unwrap();
+                    match m.get(1).unwrap().as_str() {
+                        "L" => {
+                            captures
+                                .push((pattern.start()..pattern.end(), FeatureType::Index(idx)));
+                        }
+                        "L?" => {
+                            required_indices.push(idx);
+                            captures
+                                .push((pattern.start()..pattern.end(), FeatureType::Index(idx)));
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                left_parsed_templates.push(ParsedTemplate {
+                    raw_template,
+                    required_indices,
+                    captures,
+                });
+            }
+            {
+                let raw_template = right_template.to_string();
+                let mut required_indices = vec![];
+                let mut captures = vec![];
+                for m in right_feature_pattern.captures_iter(&raw_template) {
+                    let pattern = m.get(0).unwrap();
+                    let idx: usize = m.get(2).unwrap().as_str().parse().unwrap();
+                    match m.get(1).unwrap().as_str() {
+                        "R" => {
+                            captures
+                                .push((pattern.start()..pattern.end(), FeatureType::Index(idx)));
+                        }
+                        "R?" => {
+                            required_indices.push(idx);
+                            captures
+                                .push((pattern.start()..pattern.end(), FeatureType::Index(idx)));
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                right_parsed_templates.push(ParsedTemplate {
+                    raw_template,
+                    required_indices,
+                    captures,
+                });
+            }
+        }
+
+        Self {
+            unigram_feature_ids: HashMap::new(),
+            left_feature_ids: HashMap::new(),
+            right_feature_ids: HashMap::new(),
+            unigram_templates: unigram_parsed_templates,
+            left_templates: left_parsed_templates,
+            right_templates: right_parsed_templates,
+        }
+    }
+
+    #[allow(unused)]
+    fn extract_feature_ids<S>(
```

Comment:
```suggestion
    /// Inserts feature patterns matched to the input templates in the hash map,
    /// while incrementally assigning new feature ids.
    /// Returns a sequence of ids of found features.
    #[allow(unused)]
    fn extract_feature_ids<S>(
```
Could you add the comment (if it's not wrong) because I needed to understand this function to review this code?